lib/Transforms/ConvertSecretExtractToStaticExtract/ConvertSecretExtractToStaticExtract.cpp

```diff
@@ -0,0 +1,212 @@
+    // transformation
+    if (!indexSecretnessLattice ||
+        !indexSecretnessLattice->getValue().isInitialized()) {
+      extractOp->emitWarning()
+          << "Secretness for tensor.extract index has not been set. "
+             "Conservatively, the transformation will be applied:";
+    }
+
+    auto indexSecretness = indexSecretnessLattice->getValue();
+
+    bool isIndexSecret =
+        !indexSecretness.isInitialized() ||
+        (indexSecretness.isInitialized() && indexSecretness.getSecretness());
+
+    // If index is not secret, no transformation is needed
+    if (!isIndexSecret) return failure();
+
+    ImplicitLocOpBuilder builder(extractOp->getLoc(), rewriter);
+
+    auto argType = extractOp.getResult().getType();
+
+    auto *extractOpLattice =
+        solver->lookupState<SecretnessLattice>(extractOp.getResult());
+
+    auto extractOpSecretness = extractOpLattice ? extractOpLattice->getValue() :
+                                                /*unknown*/ Secretness();
+
+    // Placeholder for extracted value
+    auto dummyValue = builder.create<arith::ConstantOp>(
+        argType, builder.getIntegerAttr(argType, 0));
+    // Set secretness for dummyValue
+    setValueToSecretness(solver, dummyValue, extractOpSecretness);
+
+    int size = extractOp.getTensor().getType().getShape().front();
+
+    SmallVector<Value> iterArgs = {dummyValue};
+    auto forOp = builder.create<affine::AffineForOp>(0, size, 1, iterArgs);
+    // Set secretness for induction variable
+    setValueToSecretness(solver, forOp.getInductionVar(), Secretness(false));
+
+    builder.setInsertionPointToStart(forOp.getBody());
+
+    // Check if the current index is equal to the secret index
+    auto cond = builder.create<arith::CmpIOp>(arith::CmpIPredicate::eq,
+                                              forOp.getInductionVar(), index);
+    setSecretnessOfResults(solver, cond->getOperands(), cond->getResults());
+
+    // Extract value at current index
+    auto newExtractOp = builder.create<tensor::ExtractOp>(
+        extractOp.getTensor(), forOp.getInductionVar());
+    setSecretnessOfResults(solver, extractOp->getOperands(),
+                           extractOp->getResults());
+
+    auto ifOp = builder.create<scf::IfOp>(
+        cond,
+        [&](OpBuilder &b, Location loc) {
+          // Yield value extracted at index
+          b.create<scf::YieldOp>(loc, newExtractOp.getResult());
+        },
+        [&](OpBuilder &b, Location loc) {
+          // Yield dummyValue
+          b.create<scf::YieldOp>(loc, forOp.getRegionIterArgs().front());
+        });
+    // TODO (#888): Secretness is valid for this region-bearing op only when
+    // index is strictly secret; secretness won't be valid if index has unknown
+    // secretness (i.e., when transformation occurs in the conservative case)
+    setSecretnessOfResults(solver, ifOp->getOperands(), ifOp->getResults());
```

Comment:
I don't think #888 is the right issue here, as `scf.if` DOES implement the `RegionBranchOpInterface`. Issue #888 is about stuff like `tensor.generate` which has regions, but no control flow interfaces.

In fact, I don't think there's even an issue here - you should be able to set the secreness directly, because it's just the join of `indexSecretness` with `tensorSecretness`. The extract op has secretness `indexSecretnes`, the `iter_arg` should also have `indexSecretness` and finally the condition also has `indexSecretness`. So, once more, we should be able to use the direct setting method rather than re-computing them.

Reply:
Not sure if I understand why the iter_arg takes indexSecretness, can you please clarify?

Reply:
You're right `iter_arg` should be `indexSecretness` combined with `tensorSecretness`.

Reply:
Since "join" does something different than what we need, I'd say we extract the "combining secretness" logic into a function in the Secretness class, so we can use it in situations like this.